### PR TITLE
feat: Added Recursive Bubble Sort 

### PR DESCRIPTION
**Pull request by AdityaPrakash-26**
#### Description of Change

I recently contributed to [Algorithms-Explanation](https://github.com/TheAlgorithms/Algorithms-Explanation) repository and added the explanation for `Recursive Bubble Sort`. I further decided to add implementation for the same and used the existing `bubble_sort.cpp` as a reference to write my code.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Recursive implementation of `bubble_sort.cpp`

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1487"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1487/checks?check_run_id=2441052649) warnings.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Documentation is not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md) yet.
You can also refer to the [Doxygen Guidelines](https://www.doxygen.nl/manual/docblocks.html) for how to document your code. 🙂

**Comment by AdityaPrakash-26** [comment]
Is this better?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for how to make your file/algorithm according to our [guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).

**Comment by AdityaPrakash-26** [comment]
Can you please tell me the specific area where my documentation is lacking? I used the iterative [`bubble_sort.cpp`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/bubble_sort.cpp) as a reference for my code, and my code structure and documentation are very similar to it.

**Comment by Panquesito7** [comment]
`bubble_sort.cpp` is kinda old and doesn't conform too much to our guidelines. As I mentioned above, please check the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for how to make your code as per our [guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). Thank you. 🙂

**Comment by AdityaPrakash-26** [comment]
Okay! Will do. Sorry for the confusion 😅

File: sorting/recursive_bubble_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+We need to traverse the array `n * (n-1)` times. However, if the entire array is
+already sorted, then we need to traverse it only once. Hence, O(n) is the best case
+complexity
+*/
+
+#include <cassert>   // for assert
+#include <iostream>  // for io operations
+#include <vector>    // for std::vector
+#include <array>     // for std::array
+#include <algorithm> // for std::is_sorted
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:54]:
```suggestion
#include <cassert>   /// for assert
#include <iostream>  /// for IO operations
#include <vector>    // for std::vector
#include <array>     /// for std::array
#include <algorithm> /// for std::is_sorted
```

File: sorting/recursive_bubble_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ * @author Aditya Prakash (https://adityaprakash.tech)
+ * @brief This is an implementation of recursive version of Bubble sort
+algorithm
+ *
+ * @details
+ * The working principle of the Bubble sort algorithm:
+
+Bubble sort is a simple sorting algorithm used to rearrange a set of elements in
+ascending or descending order. Bubble sort gets its name from the fact that data
+"bubbles" to the top of the dataset.
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:12]:
```suggestion
 * Bubble sort is a simple sorting algorithm used to rearrange a set of ascending or descending order elements.
 * Bubble sort gets its name from the fact that data "bubbles" to the top of the dataset.
```

File: sorting/recursive_bubble_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
+/**
+ * @file recursive_bubble_sort.cpp
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:2]:
```suggestion
 * @file
```

File: sorting/recursive_bubble_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ * @author Aditya Prakash (https://adityaprakash.tech)
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:3]:
```suggestion
 * @author [Aditya Prakash](https://adityaprakash.tech)
```

File: sorting/recursive_bubble_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ * @brief This is an implementation of recursive version of Bubble sort
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:4]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: sorting/recursive_bubble_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+algorithm
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:5]:
```suggestion
 * @brief This is an implementation of a recursive version of the Bubble sort algorithm
```

File: sorting/recursive_bubble_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ * @details
+ * The working principle of the Bubble sort algorithm:
+
+Bubble sort is a simple sorting algorithm used to rearrange a set of elements in
+ascending or descending order. Bubble sort gets its name from the fact that data
+"bubbles" to the top of the dataset.
+
+ * What is Swap?
+
+To swap two numbers mean that we interchange their values.
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:16]: All text should have a `* ` at the beginning.

File: sorting/recursive_bubble_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+/**
+ * @brief main method to run the program
+ * 
+ * @return int 
+ */
+int main() { 
+    test();
+    return 0;
+}
+
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:165]:
```suggestion
```

File: sorting/recursive_bubble_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    std::cout << std::endl;
+
+}
+
+/**
+ * @brief main method to run the program
+ * 
+ * @return int 
+ */
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:160]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work.

File: sorting/recursive_bubble_sort.cpp (hunk 7 above)
Comment (AdityaPrakash-26) [reply sorting/recursive_bubble_sort.cpp]: Fixed in latest commit :)

**Comment by AdityaPrakash-26** [comment]
Hello @Panquesito7! Did you get a chance to have a look at my changes? :-)

File: sorting/recursive_bubble_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * @brief This is a function for swapping two variables. Alternatively, the 
+ * <algorithm>'s swap function could be used too. However, this is from practice
+ * 
+ * @tparam T type of data variables to be swapped
+ * @param x first variable to be swapped
+ * @param y second variable to be swapped
+ */
+template <typename T>
+void swap(T *x, T *y) {
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:69]: C++11 already includes a `swap` function ([reference](https://en.cppreference.com/w/cpp/algorithm/swap)).

File: sorting/recursive_bubble_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    //!< calling the function after we have fixed the last element
+    recursive_bubble_sort(nums, n - 1);
+}
+}   //namespace ends
+
+/**
+ * @brief function for testing the code
+ * 
+ */
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:106]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: sorting/recursive_bubble_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    std::cout << std::endl;
+
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() { 
+    test();
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:159]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/recursive_bubble_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+ * What is Swap?
+
+ * To swap two numbers mean that we interchange their values.
+ * An additional variable is required for this operation. x = 5, y = 10.
+ * We want x = 10, y = 5. Here we create the most variable to do it.
+
+ * int z;
+ * z = x;
+ * x = y;
+ * y = z;
```
[the remark this answers is not in the document]
Comment (AdityaPrakash-26) [reply sorting/recursive_bubble_sort.cpp:21]: This part is the pseudo-code. I'm assuming x & y have been passed to the function, and are integers. I can re-write this code if it is not clear.

File: sorting/recursive_bubble_sort.cpp (hunk 10 above)
Comment (AdityaPrakash-26) [reply sorting/recursive_bubble_sort.cpp:69]: Yep! I actually wanted to provide my own implementation of the swap function. If it is not recommended, I can remove it and use the built-in function.

File: .vscode/settings.json [hunk 14]
```diff
@@ -2,5 +2,6 @@
      "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
      "editor.formatOnSave": true,
      "editor.formatOnType": true,
-     "editor.formatOnPaste": true
+     "editor.formatOnPaste": true,
+     "cmake.configureOnOpen": false
```
[the remark this answers is not in the document]
Comment (AdityaPrakash-26) [reply .vscode/settings.json]: Done!

File: sorting/recursive_bubble_sort.cpp [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    sorting::recursive_bubble_sort(&arr, size);
+    assert(std::is_sorted(std::begin(arr), std::end(arr)));
+    std::cout << " Passed\n";
+    //printing the array
+    for (int i = 0; i < size; i++) {
+        std::cout << arr[i] << ", ";
+    }
+    std::cout << std::endl;
+
+    // Example 2. Creating array of doubles.
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:130]:
```suggestion
    // 2nd example. Creating an array of type `double`.
```

File: sorting/recursive_bubble_sort.cpp [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    arr.push_back(94);
+    arr.push_back(12);
+    arr.push_back(37);
+    arr.push_back(63);
+    //array populating ends
+
+    sorting::recursive_bubble_sort(&arr, size);
+    assert(std::is_sorted(std::begin(arr), std::end(arr)));
+    std::cout << " Passed\n";
+    //printing the array
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:124]:
```suggestion
    // printing the array
```

File: sorting/recursive_bubble_sort.cpp [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    const int size = 6;
+    std::vector<int> arr;
+    //populating the array
+    arr.push_back(22);
+    arr.push_back(46);
+    arr.push_back(94);
+    arr.push_back(12);
+    arr.push_back(37);
+    arr.push_back(63);
+    //array populating ends
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:119]:
```suggestion
    // array populating ends
```

File: sorting/recursive_bubble_sort.cpp [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    recursive_bubble_sort(nums, n - 1);
+}
+}   //namespace ends
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // Example 1. Creating array of int,
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:108]:
```suggestion
    // 1st example. Creating an array of type `int`.
```

File: sorting/recursive_bubble_sort.cpp [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+    assert(std::is_sorted(std::begin(arr), std::end(arr)));
+    std::cout << " Passed\n";
+    //printing the array
+    for (int i = 0; i < size; i++) {
+        std::cout << arr[i] << ", ";
+    }
+    std::cout << std::endl;
+
+    // Example 2. Creating array of doubles.
+    std::cout << "Test 2 using doubles\n";
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:131]:
```suggestion
    std::cout << "2nd test using doubles\n";
```

File: sorting/recursive_bubble_sort.cpp [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    double_arr.push_back(62.7);
+    double_arr.push_back(12.2);
+    double_arr.push_back(43.6);
+    double_arr.push_back(74.1);
+    double_arr.push_back(57.9);
+    //array populating ends
+
+    sorting::recursive_bubble_sort(&double_arr, size);
+    assert(std::is_sorted(std::begin(double_arr), std::end(double_arr)));
+    std::cout << " Passed\n";
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:145]:
```suggestion
    std::cout << " 2nd test passed!\n";
```

File: sorting/recursive_bubble_sort.cpp [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+    arr.push_back(46);
+    arr.push_back(94);
+    arr.push_back(12);
+    arr.push_back(37);
+    arr.push_back(63);
+    //array populating ends
+
+    sorting::recursive_bubble_sort(&arr, size);
+    assert(std::is_sorted(std::begin(arr), std::end(arr)));
+    std::cout << " Passed\n";
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:123]:
```suggestion
    std::cout << " 1st test passed!\n";
```

File: sorting/recursive_bubble_sort.cpp [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+}
+}   //namespace ends
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // Example 1. Creating array of int,
+    std::cout << "Test 1 using ints\n";
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:109]:
```suggestion
    std::cout << "1st test using `int`\n";
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Amazing! Just missing a few small tweaks.

File: sorting/recursive_bubble_sort.cpp (hunk 10 above)
Comment (AdityaPrakash-26) [reply sorting/recursive_bubble_sort.cpp:69]: Okay! I will remove this function :)

**Comment by AdityaPrakash-26** [comment]
Done! :D

File: sorting/recursive_bubble_sort.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Panquesito7) [reply sorting/recursive_bubble_sort.cpp:23]: That would be great, thank you. 👍

File: sorting/recursive_bubble_sort.cpp [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+ * to the function which is then dereferenced, so that the changes are
+ * reflected in the original vector. It also accepts a second parameter of
+ * type `int` and name `n`, which is the size of the array.
+ * 
+ * @tparam T type of data variables in the array
+ * @param nums our array of elements.
+ * @param n size of the array
+ */
+template <typename T>
+void recursive_bubble_sort(std::vector<T> *nums, int n) {
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:73]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

File: sorting/recursive_bubble_sort.cpp [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+    for (int i = 0; i < size; i++) {
+        std::cout << arr[i] << ", ";
+    }
+    std::cout << std::endl;
+
+    // 2nd example. Creating an array of type `double`.
+    std::cout << "2nd test using doubles\n";
+    std::vector<double> double_arr;
+
+    //populating the array
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:121]:
```suggestion
    // populating the array
```

File: sorting/recursive_bubble_sort.cpp [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // 1st example. Creating an array of type `int`.
+    std::cout << "1st test using `int`\n";
+    const int size = 6;
+    std::vector<int> arr;
+    //populating the array
```
Comment (Panquesito7) [inline sorting/recursive_bubble_sort.cpp:99]:
```suggestion
    // populating the array
```

**Review by Panquesito7 — COMMENTED** [review]
A few missing tweaks and done! 😄

**Comment by AdityaPrakash-26** [comment]
I have made the changes that you requested! :D

**Review by Panquesito7 — APPROVED** [review]
Great work, @AdityaPrakash-26! LGTM. You've made your code very nice. 👍 🎉

**Comment by AdityaPrakash-26** [comment]
Thanks! I learnt a lot from this pull request, and I hope to make some more in the future! :)

**Review by ayaankhan98 — APPROVED** [review]
Good work!

**Pull request merged**